apis/bigqueryconnection/v1beta1/connection_reference.go

```diff
@@ -28,49 +29,47 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// NewBigQueryConnectionConnectionRef builds a BigQueryConnectionConnectionRef from the ConfigConnector BigQueryConnectionConnection object.
-func NewBigQueryConnectionConnectionRef(ctx context.Context, reader client.Reader, obj *BigQueryConnectionConnection) (*BigQueryConnectionConnectionRef, error) {
-	id := &BigQueryConnectionConnectionRef{}
+type ConnectionIdentity struct {
+	Parent       *parent.ProjectAndLocationParent
+	ConnectionID string
+}
 
-	projectRef, err := refsv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
-	if err != nil {
+func (c *ConnectionIdentity) String() string {
+	return c.Parent.String() + "/connections/" + c.ConnectionID
+}
+
+// NewBigQueryConnectionConnectionRef builds a BigQueryConnectionConnectionRef from the ConfigConnector BigQueryConnectionConnection object.
+func NewConnectionIdentity(ctx context.Context, reader client.Reader, obj *BigQueryConnectionConnection) (*ConnectionIdentity, error) {
+	desiredParent := new(parent.ProjectAndLocationParent)
+	if err := obj.Spec.ProjectAndLocationRef.Build(ctx, reader, obj.GetNamespace(), desiredParent); err != nil {
 		return nil, err
 	}
-	projectID := projectRef.ProjectID
-	if projectID == "" {
-		return nil, fmt.Errorf("cannot resolve project")
-	}
-	// Get location
-	location := obj.Spec.Location
 
 	// Get desired connection ID from spec
 	desiredID := direct.ValueOf(obj.Spec.ResourceID)
 
 	// Validate status.externalRef
 	externalRef := direct.ValueOf(obj.Status.ExternalRef)
 	if externalRef != "" {
-		actualProject, actualLocation, actualID, err := parseExternal(externalRef)
+		actualParent, actualID, err := ParseConnectionExternal(externalRef)
 		if err != nil {
 			return nil, err
 		}
-		// Validate spec parent and resourceID field if the resource is already reconcilied with a GCP Connection resource.
-		if projectID != actualProject {
-			return nil, fmt.Errorf("spec.projectRef changed, expect %s, got %s", projectID, actualProject)
-		}
-		if location != actualLocation {
-			return nil, fmt.Errorf("spec.location changed, expect %s, got %s", location, actualLocation)
+		if err := desiredParent.MatchActual(actualParent); err != nil {
```

Comment:
Suggest just comparing `String()` of the ID

Reply:
This `MatchActual` is more for KCC users. 
My main concern about comparing `String` is that it cannot give users enough information to help them fix their bad config.
For example, the parent is build from `spec.ProjectRef.External` and `spec.Location` (for some other resources it may also have `spec. FooRef.External` `spec.BarRef.External`), it is unclear for users to know which field is misconfigured by just seeing something like`expected format projects/<projectID>/location/<lcoation>/foos/<fooid>/bars/<bar>, got ....`

Reply:
Well I think I would make the message "Can't change object identity; existing identity is %q, new identity is %q" so it would be pretty easy to figure out what has changed, but ... I can send a PR on top of this - let's get this in!